zbus/src/connection/mod.rs

```diff
@@ -1370,6 +1402,112 @@ mod tests {
         let name_has_owner = dbus.name_has_owner(name.try_into().unwrap()).await.unwrap();
         assert!(!name_has_owner);
     }
+
+    #[tokio::test(start_paused = true)]
+    #[timeout(15000)]
+    async fn test_graceful_shutdown() {
+        // If we have the only reference, it should complete immediately.
+        let connection = Connection::session().await.unwrap();
+        let shutdown = connection.graceful_shutdown();
+        tokio::select! {
+            biased;
+            _ = shutdown => {},
+            _ = std::future::ready(()) => {
+                panic!("Graceful shutdown did not immediately complete");
+            }
+        }
+
+        // If we have a second reference, it should wait until we drop it.
+        let connection = Connection::session().await.unwrap();
+        let clone = connection.clone();
+        let mut shutdown = pin!(connection.graceful_shutdown());
+        tokio::select! {
+            _ = tokio::time::sleep(Duration::from_secs(u64::MAX)) => {},
+            _ = &mut shutdown => {
+                panic!("Graceful shutdown unexpectedly completed");
+            }
```

Comment:
I'm not following. Only 2 futures and you don't want either of them to complete in our lifetimes? :smile:

Reply:
Thanks so much for explaining in detail. I appreciate you trying so hard to be very precise. :+1: I learnt something new today. :) I'm not sure we've a similar API/mechanism available in the smol-rs/async-io world, so this part will have to be tokio-specific but that's no biggie since the main part can be backend-agnostic. However, please add a short comment (perhaps above the `sleep` arm of the `select`). Something like "Tokio will advance the time to ensure this future get resolved: https://docs.rs/tokio/latest/tokio/time/fn.pause.html#auto-advance" should suffice.